ccmbioinfo/lfs
# CARDS-1783: Make email scheduling more flexible

The PR removes the `InitialNotificationsTask` and `ReminderNotificationsTask` scheduled tasks and provides, as a replacement, an OSGi component that can be configured and instantiated to provide emails for upcoming appointments. The functionality of the `InitialNotificationsTask` and `ReminderNotificationsTask` scheduled tasks is preserved by the `InitialNotificationsTask` and `ReminderNotificationsTask` configurations for `io.uhndata.cards.proms.emailnotifications.UpcomingNotificationsFactory` that is loaded upon repository initialization.

To test:

- Build this PR with `mvn clean install`
- Follow the instructions in https://github.com/data-team-uhn/cards/blob/dev/Utilities/Development/EmailServer/README.md to setup a testing mail server and client to see the emails sent by CARDS. Follow the _Using (with `./start_cards.sh`)_ instructions as opposed to the _Using (with Docker Compose)_ instructions.
- Open a new terminal in `Utilities/Development` and start the mock Torch server by running `nodejs mock_torch.js --appointment-time-hours-from-now=24`.
- Open a new terminal in `Utilities/Development` and start the mock Slack server by running `nodejs mock_slack.js`
- Start CARDS in _Cards4Proms_ mode with the command `PATIENT_NOTIFICATION_FROM_ADDRESS=datapro@uhn.ca PATIENT_NOTIFICATION_FROM_NAME=DATAPRO NIGHTLY_NOTIFICATIONS_SCHEDULE="0 * * * * ? *" CARDS_HOST_AND_PORT='localhost:8080' NIGHTLY_SLACK_NOTIFICATIONS_SCHEDULE="0 * * * * ? *" SLACK_PERFORMANCE_URL="http://localhost:8000/" ./start_cards.sh --dev --project cards4proms -V emailnotifications.smtps.checkserveridentity=false`.
- Visit `http://localhost:8080` and login as `admin`:`admin`.
- Visit `http://localhost:8080/system/console/configMgr` and add a new _PROMs Import_ configuration with the following values:
  - `Name`: `LocalNodeJS`
  - `Import schedule`: `0 0 0 * * ? *`
  - `days to query`: `3`
  - `Endpoint URL`: `http://localhost:8011/`
  - `Authentication URL`: (leave blank)
  - `Vault token`: (leave blank)
  - `Clinic names`: `6012-HC-Congenital Cardiac`
  - `Provider names`: `SomeParticipantEID`
  - `Allowed provider roles`: `ATND`
  - `Vault role name`: (leave blank)
  - `Dates to query`: (leave blank)
- Trigger the import by visiting `http://localhost:8080/Subjects.importTorch?config=LocalNodeJS`
- Verify that the `12345 / AppointmentOneFhirID-Cardio` appointment was created.
- You should now receive an email with the subject _"Reminder: You have 24 hours left to complete your pre-appointment questions"_ once every minute.
- Visit `http://localhost:8080/system/console/configMgr` once again and create a new _Upcoming appointment notification_ with the following values:
  - `Name`: `TestNotifications`
  - `Metric Name`: `Test Notification emails sent`
  - `Clinic Mapping Path`: `/Proms/ClinicMapping/853703519`
  - `Email Subject Line`: `Email Notification System Test`
  - `Plaintext Email Template JCR Path`: `/apps/cards/proms/clinics/Cardio/mailTemplates/24h.txt`
  - `HTML Email Template JCR Path`: `/apps/cards/proms/clinics/Cardio/mailTemplates/24h.html`
  - `Days to the visit`: `1`
- Every minute you should receive an upcoming appointment notification email with the subject line `Email Notification System Test`.
- On the mock Slack server, you should also be receiving, once every minute, performance notifications that include an incrementing count of `Test Notification emails sent` and `Reminder Emails Sent`. It should also display `1` for the `Total` number of `Imported Appointments`.
- Repeat the above steps but, this time, specify an appointment 72 hours from now instead of 24 hours when starting the mock Torch server (ie. `nodejs mock_torch.js --appointment-time-hours-from-now=72`). Verify that you are receiving an email with the subject line _"Welcome to DATAPRO: Answer your Pre-Appointment Questions"_ once every minute. Also verify that the `Initial Emails Sent` count in the Slack notifications is auto-incrementing.
